app-kotlin/src/main/kotlin/au/com/gridstone/trainingkotlin/screens/home/HomeViewController.kt

```diff
@@ -4,57 +4,152 @@ import android.os.Bundle
+            )
+          }
+        }
+      }
+    }
+  }
+}
 
-  override fun onAttach(view: View) {
+@Composable
+fun PokemonHomeContent(viewModel: HomeViewModel, onSelectPokemon: (Pokemon) -> Unit) {
 
-    val presenter = HomeViewPresenter(view)
+  val state: HomeViewState by viewModel.states.collectAsState()
 
-    launch {
-      presenter.events.collect { event: HomeViewEvents ->
-        when (event) {
-          is HomeViewEvents.Refresh -> launch(Dispatchers.IO) { viewModel.getAllPokemon() }
-          is HomeViewEvents.ItemClick -> router.pushController(
-            RouterTransaction.with(DetailViewController(event.pokemon))
-              .pushChangeHandler(HorizontalChangeHandler())
-              .popChangeHandler(HorizontalChangeHandler())
-          )
-        }
+  Scaffold(
+    topBar = {
+      TopAppBar {
+        Text(
+          text = stringResource(id = R.string.pokemon),
+          style = MaterialTheme.typography.h6,
+        )
       }
     }
-
-    launch {
-      viewModel.states
-        .collect { state: HomeViewState ->
-          presenter.display(state)
+  ) { padding ->
+    SwipeRefresh(
+      modifier = Modifier
+        .fillMaxSize()
+        .padding(padding),
+      state = rememberSwipeRefreshState(isRefreshing = state == Loading),
+      onRefresh = { with(viewModel) { launch { getAllPokemon() } } }
+    ) {
+      Column(verticalArrangement = Arrangement.SpaceEvenly,
+        horizontalAlignment = Alignment.CenterHorizontally,
+        modifier = Modifier.fillMaxSize()) {
+        when (val _state: HomeViewState = state) {
+          Loading -> {
+            LoadingView()
+          }
+          Failed -> {
+            FailedView()
+          }
+          is Success -> {
+            PokemonList(_state.results.results, onSelectPokemon = onSelectPokemon)
+          }
         }
+      }
     }
   }
+}
+
+@Composable
+fun FailedView() {
+  Column(
+    modifier = Modifier.wrapContentSize(),
+    horizontalAlignment = Alignment.CenterHorizontally,
+  ) {
+    Image(
+      painter = painterResource(id = R.drawable.ic_error_round),
+      contentDescription = "Error",
+      modifier = Modifier
+        .size(50.dp),
+      alignment = Alignment.Center,
+    )
+    Text(text = "Something went wrong")
+  }
+}
 
-  override fun onDetach(view: View) {
-    coroutineContext.cancelChildren()
+@Composable
+fun LoadingView() {
+  CircularProgressIndicator(progress = 0.8f,
+    modifier = Modifier.padding(8.dp),
+    color = Companion.DarkGray,
+    strokeWidth = 2.dp)
+}
+
+@Composable
+fun PokemonList(pokemonList: List<Pokemon>, onSelectPokemon: (Pokemon) -> Unit) {
+  LazyColumn(
+    contentPadding = PaddingValues(horizontal = 8.dp, vertical = 8.dp)
+  ) {
+    items(
+      items = pokemonList,
+      itemContent = {
+        Surface(modifier = Modifier.clickable { onSelectPokemon(it) }) {
+          PokemonListItem(pokemon = it)
+        }
+      }
```

Comment:
given that this is a slot, you can lift this out of the function invocation 
```kotlin
items(pokemonList) {
  Surface(modifier = Modifier.clickable { onSelectPokemon(it) }) {
    PokemonListItem(pokemon = it)
  }
}
```

Reply:
sure